File: atm-8/ATM.Logic/AbstractIE.cs
using System;
using System.Collections.Generic;
using ATM.Domain;

namespace ATM.Logic
{
    public abstract class AbstractIE
    {
        protected void CheckRules()
        {
            List<Object> rules = GetAllRules();
            foreach (Object rule in rules)
            {
                if (CheckRuleConditions(GetRuleOperator(rule), GetRuleConditions(rule)))
                {
                    ExecuteConsequences(GetRuleConsequences(rule));
                }
            }
        }

        protected abstract List<Object> GetAllRules();

        protected abstract RuleOperator GetRuleOperator(object rule);

        protected abstract List<Object> GetRuleConditions(Object rule);

        protected abstract List<Object> GetRuleConsequences(Object rule);

        protected abstract String GetConditionExpression(Object condition);

        protected abstract String GetConsequenceExpression(Object consequence);

        /**
         * The returned parameter can be a fact object, or a String.
         * 
         * @param consequence
         * @return
         */
        protected abstract Object GetConsequenceParameter(Object consequence);

        protected abstract Object GetConditionParameter(Object condition);

        private void ExecuteConsequences(List<Object> consequences)
        {
            ClearRuntimeFactStorage();
            foreach (Object consequence in consequences)
            {
                if (GetConsequenceExpression(consequence) == (ClauseOperators.Print.ToString()))
                {
                    Console.WriteLine(GetConsequenceParameter(consequence));
                }
                else if (GetConsequenceExpression(consequence) == (ClauseOperators.AssertFact.ToString()))
                {
                    Object fact = (Object)GetConsequenceParameter(consequence);
                    EvaluateSlots(fact);
                    AssertFact(fact);
                }
            }
        }

        /**
         * Completely clean runtime storage.
         */
        protected abstract void ClearRuntimeFactStorage();

        protected abstract List<Object> GetAllFactsInRuntimeStorage();

        /**
         * Put new fact to runtime storage.
         * 
         * @param fact
         */
        protected abstract void AssertFact(Object fact);

        private void EvaluateSlots(Object fact)
        {
            foreach (Object property in GetFactProperties(fact))
            {
                if (PropertyRequiresInputFromUser(property))
                {
                    try
                    {
                        Console.WriteLine("Please enter " + GetFactPropertyName(property) + ": ");
                        String s = Console.ReadLine();
                        Type type = Type.GetType(GetFactPropertyType(property), true);
                        var instance = Activator.CreateInstance(type, s);
                        SetFactPropertyValue(property, instance);
                    }
                    catch (Exception ex)
                    {
                        Console.Error.WriteLine(ex);
                    }

                }
            }
        }

        protected abstract List<Object> GetFactProperties(Object fact);

        protected abstract String GetFactName(Object fact);

        protected abstract bool PropertyRequiresInputFromUser(Object property);

        /**
         * This method MUST return a valid fully qualified java class name!
         * 
         * @param property
         * @return
         */
        protected abstract String GetFactPropertyType(Object property);

        protected abstract void SetFactPropertyValue(Object property, Object value);

        protected abstract String GetFactPropertyName(Object property);

        protected abstract void SetConditionParameter(Object condition, Object parameter);

        private bool CheckRuleConditions(RuleOperator ruleOp, List<Object> conditions)
        {
            bool conditionMet = true;
            IList<Object> retreivedFacts = new List<Object>();
            foreach (Object clause in conditions)
            {
                conditionMet = conditionMet & EvaluateClause(clause, retreivedFacts);
            }
            return conditionMet;
        }

        private bool EvaluateClause(Object clause, IList<Object> retreivedFacts)
        {
            if (GetConditionExpression(clause) == (ClauseOperators.Exists.ToString()))
            {
                if (GetConditionParameter(clause) == null)
                {
                    SetConditionParameter(clause, retreivedFacts[0]);
                }
                return Exists(GetConditionParameter(clause));
            }
            else if (GetConditionExpression(clause) == (ClauseOperators.Take.ToString()))
            {
                Object fact = GetFirstFact(GetAllFactsInRuntimeStorage(), GetConditionParameter(clause).ToString());
                if (fact != null)
                {
                    retreivedFacts.Add(fact);
                    return true;
                }
            }

            return false;
        }

        private Object GetFirstFact(List<Object> inferredFacts, String factName)
        {
            Object fact = null;
            int i = 0;
            while (fact == null && i < inferredFacts.Count)
            {
                if (GetFactName(inferredFacts[i]) == (factName))
                {
                    fact = inferredFacts[i];
                }
                i++;
            }
            return fact;
        }

        protected abstract List<Object> GetAllFacts();

        protected abstract int GetNrOfFactProperties(Object fact);

        protected abstract Object GetFactPropertyOnPosition(Object fact, int position);

        protected abstract bool IsFactPropertyMandatory(Object property);

        protected abstract Object GetFactPropertyValue(Object property);

        private bool Exists(Object factToCheck)
        {
            List<Object> facts = GetAllFacts();
            foreach (Object fact in facts)
            {
                bool factExists = true;

                factExists = factExists && GetFactName(fact) == (GetFactName(factToCheck));
                factExists = factExists && (GetNrOfFactProperties(fact) == GetNrOfFactProperties(factToCheck));

                for (int i = 0; i < GetNrOfFactProperties(fact); i++)
                {
                    Object slot = GetFactPropertyOnPosition(fact, i);
                    if (IsFactPropertyMandatory(slot))
                    {
                        Object slotToCheck = GetFactPropertyOnPosition(factToCheck, i);

                        factExists = factExists && GetFactPropertyName(slot) == (GetFactPropertyName(slotToCheck));
                        factExists = factExists && GetFactPropertyType(slot) == (GetFactPropertyType(slotToCheck));
                        factExists = factExists && GetFactPropertyValue(slot) == (GetFactPropertyValue(slotToCheck));
                    }
                }

                if (factExists)
                {
                    return true;
                }
            }

            return false;
        }
    }
}

File: atm-10/src/inference/AbstractIE.java
package inference;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.lang.reflect.Constructor;
import java.util.ArrayList;
import java.util.List;

import inference.model.Condition;
import inference.model.Fact;
import inference.model.Property;
import inference.model.Rule;
import inference.model.Consequence;

public abstract class AbstractIE {

	protected void checkRules() {
		List<Rule> rules = getAllRules();
		for (Rule rule : rules) {
			if (checkRuleConditions(getRuleOperator(rule), getRuleConditions(rule))) {
				executeConsequences(getRuleConsequences(rule));
			}
		}
	}

	protected abstract List<Rule> getAllRules();

	protected abstract RuleOperator getRuleOperator(Rule rule);

	protected abstract List<Condition> getRuleConditions(Rule rule);

	protected abstract List<Consequence> getRuleConsequences(Rule rule);

	protected abstract String getConditionExpression(Condition condition);

	protected abstract String getConsequenceExpression(Consequence consequence);

	/**
	 * The returned parameter can be a fact object, or a String.
	 * 
	 * @param consequence
	 * @return
	 */
	protected abstract Fact getConsequenceParameter(Consequence consequence);

	protected abstract Fact getConditionParameter(Condition condition);

	private void executeConsequences(List<Consequence> consequences) {
		clearRuntimeFactStorage();
		for (Consequence consequence : consequences) {
			if (getConsequenceExpression(consequence).equals(ClauseOperators.print.name())) {
				System.out.println(getConsequenceParameter(consequence));
			} else if (getConsequenceExpression(consequence).equals(ClauseOperators.assertFact.name())) {
				Fact fact = getConsequenceParameter(consequence);
				evaluateSlots(fact);
				assertFact(fact);
			}
		}
	}

	/**
	 * Completely clean runtime storage.
	 */
	protected abstract void clearRuntimeFactStorage();

	protected abstract List<Fact> getAllFactsInRuntimeStorage();

	/**
	 * Put new fact to runtime storage.
	 * 
	 * @param fact
	 */
	protected abstract void assertFact(Fact fact);

	private void evaluateSlots(Fact fact) {
		for (Property property : getFactProperties(fact)) {
			if (propertyRequiresInputFromUser(property)) {
				try {
					System.out.println("Please enter " + getFactPropertyName(property) + ": ");
					BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
					String s = br.readLine();
					Constructor<?> ctor = Class.forName(getFactPropertyType(property)).getConstructor(String.class);
					//???
					setFactPropertyValue(property, (String) ctor.newInstance(s));
				} catch (Exception ex) {
					ex.printStackTrace();
				}

			}
		}
	}

	protected abstract List<Property> getFactProperties(Fact fact);

	protected abstract String getFactName(Fact fact);

	protected abstract boolean propertyRequiresInputFromUser(Property property);

	/**
	 * This method MUST return a valid fully qualified java class name!
	 * 
	 * @param property
	 * @return
	 */
	protected abstract String getFactPropertyType(Property property);

	protected abstract void setFactPropertyValue(Property property, String value);

	protected abstract String getFactPropertyName(Property property);

	protected abstract void setConditionParameter(Condition condition, Fact parameter);

	private boolean checkRuleConditions(RuleOperator operator, List<Condition> conditions) {
		boolean conditionMet = true;
		List<Fact> retreivedFacts = new ArrayList<>();
		for (Condition clause : conditions) {
			conditionMet = conditionMet & evaluateClause(clause, retreivedFacts);
		}
		return conditionMet;
	}

	private boolean evaluateClause(Condition clause, List<Fact> retreivedFacts) {
		if (getConditionExpression(clause).equals(ClauseOperators.exists.name())) {
			if (getConditionParameter(clause) == null) {
				setConditionParameter(clause, retreivedFacts.get(0));
			}
			return exists((Fact)getConditionParameter(clause));
		} else if (getConditionExpression(clause).equals(ClauseOperators.take.name())) {
			Fact fact = getFirstFact(getAllFactsInRuntimeStorage(), getConditionParameter(clause).toString());
			if (fact != null) {
				retreivedFacts.add(fact);
				return true;
			}
		}

		return false;
	}

	private Fact getFirstFact(List<Fact> inferredFacts, String factName) {
		Fact fact = null;
		int i = 0;
		while (fact == null && i < inferredFacts.size()) {
			if (getFactName(inferredFacts.get(i)).equals(factName)) {
				fact = inferredFacts.get(i);
			}
			i++;
		}
		return fact;
	}

	protected abstract List<Fact> getAllFacts();

	protected abstract int getNrOfFactProperties(Fact fact);

	protected abstract Property getFactPropertyOnPosition(Fact fact, int position);

	protected abstract boolean isFactPropertyMandatory(Property property);

	protected abstract String getFactPropertyValue(Property property);

	private boolean exists(Fact factToCheck) {
		List<Fact> facts = getAllFacts();
		for (Fact fact : facts) {
			boolean factExists = true;

			factExists = factExists && getFactName(fact).equals(getFactName(factToCheck));
			factExists = factExists && (getNrOfFactProperties(fact) == getNrOfFactProperties(factToCheck));

			for (int i = 0; i < getNrOfFactProperties(fact); i++) {
				Property slot = getFactPropertyOnPosition(factToCheck, i);
				if (isFactPropertyMandatory(slot)) {
					Property slotToCheck = getFactPropertyOnPosition(factToCheck, i);

					factExists = factExists && getFactPropertyName(slot).equals(getFactPropertyName(slotToCheck));
					factExists = factExists && getFactPropertyType(slot).equals(getFactPropertyType(slotToCheck));
					factExists = factExists && getFactPropertyValue(slot).equals(getFactPropertyValue(slotToCheck));
				}
			}

			if (factExists) {
				return true;
			}
		}

		return false;
	}

}
